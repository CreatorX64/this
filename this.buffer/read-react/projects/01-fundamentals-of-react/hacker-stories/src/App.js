import { useState, useEffect, useRef, useCallback } from "react";

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = useState(localStorage.getItem(key) ?? initialState);

  useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);

  return [value, setValue];
};

export const App = () => {
  const stories = [
    {
      objectId: 0,
      title: "React",
      url: "https://reactjs.org",
      author: "Jordan Walke",
      numComments: 3,
      points: 4
    },
    {
      objectId: 1,
      title: "Redux",
      url: "https://redux.js.org",
      author: "Dan Abramov, Andrew Clark",
      numComments: 2,
      points: 5
    }
  ];

  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");

  const handleSearch = (event) => {
    setSearchTerm(event.target.value);
  };

  const searchedStories = stories.filter((story) =>
    story.title.toLowerCase().includes(searchTerm.toLowerCase())
  );

  return (
    <div>
      <h1>My Hacker Stories</h1>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={handleSearch}
        isFocused
      >
        <strong>Search:</strong>
      </InputWithLabel>
      <hr />
      <List list={searchedStories} />
    </div>
  );
};

const InputWithLabel = ({
  children,
  id,
  value,
  type = "text",
  onInputChange,
  isFocused
}) => {
  const inputRef = useRef();

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <>
      <label htmlFor={id}>{children}</label>
      &nbsp;
      <input
        ref={inputRef}
        type={type}
        id={id}
        value={value}
        onChange={onInputChange}
        // autoFocus={isFocused}
      />
    </>
  );
};

const List = ({ list }) => (
  <ul>
    {list.map((item) => (
      <Item key={item.objectId} item={item} />
    ))}
  </ul>
);

const Item = ({ item }) => (
  <li>
    <span>
      <a href={item.url}>{item.title}</a>
    </span>{" "}
    <span>{item.author}</span> <span>{item.numComments}</span>{" "}
    <span>{item.points}</span>
  </li>
);
