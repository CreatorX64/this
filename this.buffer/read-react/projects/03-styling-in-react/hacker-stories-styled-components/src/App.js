import { useState, useReducer, useEffect, useRef, useCallback } from "react";
import axios from "axios";
import styled from "styled-components";

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

// Action types

const REMOVE_STORY = "REMOVE_STORY";
const STORIES_FETCH_INIT = "STORIES_FETCH_INIT";
const STORIES_FETCH_SUCCESS = "STORIES_FETCH_SUCCESS";
const STORIES_FETCH_FAILURE = "STORIES_FETCH_FAILURE";

// Styled components

const StyledContainer = styled.div`
  min-height: 100vh;
  padding: 20px;
  color: #171212;
  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);
`;

const StyledHeadlinePrimary = styled.h1`
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
`;

const StyledItem = styled.li`
  display: flex;
  align-items: center;
  padding-bottom: 5px;
`;

const StyledColumn = styled.span`
  overflow: hidden;
  padding: 0 5px;
  white-space: nowrap;
  text-overflow: ellipsis;

  a {
    color: inherit;
  }

  width: ${(props) => props.width};
`;

const StyledButton = styled.button`
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  background: transparent;
  transition: all 0.1s ease-in;

  &:hover {
    color: #fff;
    background: #171212;
  }
`;

const StyledButtonSmall = styled(StyledButton)`
  padding: 5px;
`;

const StyledButtonLarge = styled(StyledButton)`
  padding: 10px;
`;

const StyledSearchForm = styled.form`
  display: flex;
  align-items: baseline;
  gap: 8px;
  padding: 10px 0 20px;
`;

const StyledLabel = styled.label`
  padding-left: 5px;
  font-size: 24px;
`;

const StyledInput = styled.input`
  border: none;
  border-bottom: 1px solid #171212;
  font-size: 24px;
  background-color: transparent;
`;

// Reducers

const storiesReducer = (state, action) => {
  switch (action.type) {
    case STORIES_FETCH_INIT:
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case STORIES_FETCH_SUCCESS:
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case STORIES_FETCH_FAILURE:
      return {
        ...state,
        isLoading: false,
        isError: true
      };
    case REMOVE_STORY:
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        )
      };
    default:
      throw new Error();
  }
};

// Hooks

const useSemiPersistentState = (key, initialState) => {
  const [value, setValue] = useState(localStorage.getItem(key) ?? initialState);

  useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);

  return [value, setValue];
};

// Components

export const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false
  });

  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: STORIES_FETCH_INIT });

    try {
      const result = await axios.get(url);

      dispatchStories({
        type: STORIES_FETCH_SUCCESS,
        payload: result.data.hits
      });
    } catch {
      dispatchStories({ type: STORIES_FETCH_FAILURE });
    }
  }, [url]);

  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleSearchInput = (event) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event) => {
    event.preventDefault();
    setUrl(`${API_ENDPOINT}${searchTerm}`);
  };

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: REMOVE_STORY,
      payload: item
    });
  };

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Something went wrong :(</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </StyledContainer>
  );
};

const SearchForm = ({ searchTerm, onSearchInput, onSearchSubmit }) => (
  <StyledSearchForm onSubmit={onSearchSubmit}>
    <InputWithLabel
      id="search"
      value={searchTerm}
      onInputChange={onSearchInput}
      isFocused
    >
      <strong>Search:</strong>
    </InputWithLabel>

    <StyledButtonLarge
      type="submit"
      disabled={!searchTerm || !searchTerm.trim()}
    >
      Submit
    </StyledButtonLarge>
  </StyledSearchForm>
);

const InputWithLabel = ({
  children,
  id,
  value,
  type = "text",
  onInputChange,
  isFocused
}) => {
  const inputRef = useRef();

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <>
      <StyledLabel htmlFor={id}>{children}</StyledLabel>
      &nbsp;
      <StyledInput
        ref={inputRef}
        type={type}
        id={id}
        value={value}
        onChange={onInputChange}
        // autoFocus={isFocused}
      />
    </>
  );
};

const List = ({ list, onRemoveItem }) => (
  <ul>
    {list.map((item) => (
      <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
    ))}
  </ul>
);

const Item = ({ item, onRemoveItem }) => (
  <StyledItem>
    <StyledColumn style={{ width: "40%" }}>
      <a href={item.url}>{item.title}</a>
    </StyledColumn>
    <StyledColumn style={{ width: "30%" }}>{item.author}</StyledColumn>
    <StyledColumn style={{ width: "10%" }}>{item.numComments}</StyledColumn>
    <StyledColumn style={{ width: "10%" }}>{item.points}</StyledColumn>
    <StyledColumn style={{ width: "10%" }}>
      <StyledButtonSmall type="button" onClick={() => onRemoveItem(item)}>
        Dismiss
      </StyledButtonSmall>
    </StyledColumn>
  </StyledItem>
);
