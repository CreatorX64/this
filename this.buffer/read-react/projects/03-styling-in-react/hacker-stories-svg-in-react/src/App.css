.container {
  min-height: 100vh;
  padding: 20px;
  color: #171212;
  background: #83a4d4; /* fallback */
  background: linear-gradient(to left, #b6fbff, #83a4d4);
}

.headline-primary {
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
}

.item {
  display: flex;
  align-items: center;
  padding-bottom: 5px;
}

.item > span {
  overflow: hidden;
  padding: 0 5px;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.item > span > a {
  color: inherit;
}

.button {
  border: 1px solid #171212;
  padding: 5px;
  cursor: pointer;
  background: transparent;
  transition: all 0.1s ease-in;
}

.button:hover {
  color: #fff;
  background: #171212;
}

.button:hover > svg > g {
  fill: #fff;
  stroke: #fff;
}

.button_small {
  padding: 5px;
}

.button_large {
  padding: 10px;
}

.search-form {
  display: flex;
  align-items: baseline;
  gap: 8px;
  padding: 10px 0 20px;
}

.label {
  padding-left: 5px;
  font-size: 24px;
}

.input {
  border: none;
  border-bottom: 1px solid #171212;
  font-size: 24px;
  background-color: transparent;
}
