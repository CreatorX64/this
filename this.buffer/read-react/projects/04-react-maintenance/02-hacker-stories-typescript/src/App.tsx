import React, {
  useState,
  useReducer,
  useEffect,
  useRef,
  useCallback
} from "react";
import axios from "axios";
import styles from "./App.module.css";

const API_ENDPOINT = "https://hn.algolia.com/api/v1/search?query=";

type Story = {
  objectID: string;
  url: string;
  title: string;
  author: string;
  num_comments: number;
  points: number;
};

type Stories = Array<Story>;

const useSemiPersistentState = (
  key: string,
  initialState: string
): [string, (newValue: string) => void] => {
  const [value, setValue] = useState(localStorage.getItem(key) ?? initialState);

  useEffect(() => {
    localStorage.setItem(key, value);
  }, [value, key]);

  return [value, setValue];
};

type StoriesState = {
  data: Stories;
  isLoading: boolean;
  isError: boolean;
};

interface StoriesFetchInitAction {
  type: "STORIES_FETCH_INIT";
}

interface StoriesFetchSuccessAction {
  type: "STORIES_FETCH_SUCCESS";
  payload: Stories;
}

interface StoriesFetchFailureAction {
  type: "STORIES_FETCH_FAILURE";
}

interface StoriesRemoveAction {
  type: "REMOVE_STORY";
  payload: Story;
}

type StoriesAction =
  | StoriesFetchInitAction
  | StoriesFetchSuccessAction
  | StoriesFetchFailureAction
  | StoriesRemoveAction;

const storiesReducer = (state: StoriesState, action: StoriesAction) => {
  switch (action.type) {
    case "STORIES_FETCH_INIT":
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case "STORIES_FETCH_SUCCESS":
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: action.payload
      };
    case "STORIES_FETCH_FAILURE":
      return {
        ...state,
        isLoading: false,
        isError: true
      };
    case "REMOVE_STORY":
      return {
        ...state,
        data: state.data.filter(
          (story) => action.payload.objectID !== story.objectID
        )
      };
    default:
      throw new Error();
  }
};

export const App = () => {
  const [searchTerm, setSearchTerm] = useSemiPersistentState("search", "React");
  const [url, setUrl] = useState(`${API_ENDPOINT}${searchTerm}`);
  const [stories, dispatchStories] = useReducer(storiesReducer, {
    data: [],
    isLoading: false,
    isError: false
  });

  const handleFetchStories = useCallback(async () => {
    dispatchStories({ type: "STORIES_FETCH_INIT" });

    try {
      const result = await axios.get(url);

      dispatchStories({
        type: "STORIES_FETCH_SUCCESS",
        payload: result.data.hits
      });
    } catch {
      dispatchStories({ type: "STORIES_FETCH_FAILURE" });
    }
  }, [url]);

  useEffect(() => {
    handleFetchStories();
  }, [handleFetchStories]);

  const handleSearchInput = (event: React.ChangeEvent<HTMLInputElement>) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    setUrl(`${API_ENDPOINT}${searchTerm}`);
  };

  const handleRemoveStory = (item: Story) => {
    dispatchStories({
      type: "REMOVE_STORY",
      payload: item
    });
  };

  return (
    <div className={styles.container}>
      <h1 className={styles.headlinePrimary}>My Hacker Stories</h1>

      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      />

      {stories.isError && <p>Something went wrong :(</p>}

      {stories.isLoading ? (
        <p>Loading...</p>
      ) : (
        <List list={stories.data} onRemoveItem={handleRemoveStory} />
      )}
    </div>
  );
};

type SearchFormProps = {
  searchTerm: string;
  onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
  onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
};

const SearchForm = ({
  searchTerm,
  onSearchInput,
  onSearchSubmit
}: SearchFormProps) => {
  return (
    <form onSubmit={onSearchSubmit} className={styles.searchForm}>
      <InputWithLabel
        id="search"
        value={searchTerm}
        onInputChange={onSearchInput}
        isFocused
      >
        <strong>Search:</strong>
      </InputWithLabel>

      <button
        type="submit"
        disabled={!searchTerm || !searchTerm.trim()}
        className={`${styles.button} ${styles.buttonLarge}`}
      >
        Submit
      </button>
    </form>
  );
};

type InputWithLabelProps = {
  id: string;
  value: string;
  type?: string;
  onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  isFocused?: boolean;
  children: React.ReactNode;
};

const InputWithLabel = ({
  children,
  id,
  value,
  type = "text",
  onInputChange,
  isFocused
}: InputWithLabelProps) => {
  const inputRef = useRef<HTMLInputElement>(null!);

  useEffect(() => {
    if (isFocused && inputRef.current) {
      inputRef.current.focus();
    }
  }, [isFocused]);

  return (
    <>
      <label htmlFor={id} className={styles.label}>
        {children}
      </label>
      &nbsp;
      <input
        ref={inputRef}
        type={type}
        id={id}
        value={value}
        onChange={onInputChange}
        className={styles.input}
        // autoFocus={isFocused}
      />
    </>
  );
};

type ListProps = {
  list: Stories;
  onRemoveItem: (item: Story) => void;
};

const List = ({ list, onRemoveItem }: ListProps) => (
  <ul>
    {list.map((item) => (
      <Item key={item.objectID} item={item} onRemoveItem={onRemoveItem} />
    ))}
  </ul>
);

type ItemProps = {
  item: Story;
  onRemoveItem: (item: Story) => void;
};

const Item = ({ item, onRemoveItem }: ItemProps) => (
  <li className={styles.item}>
    <span style={{ width: "40%" }}>
      <a href={item.url}>{item.title}</a>
    </span>
    <span style={{ width: "30%" }}>{item.author}</span>
    <span style={{ width: "10%" }}>{item.num_comments}</span>
    <span style={{ width: "10%" }}>{item.points}</span>
    <span style={{ width: "10%" }}>
      <button
        type="button"
        className={`${styles.button} ${styles.buttonSmall}`}
        onClick={() => onRemoveItem(item)}
      >
        Dismiss
      </button>
    </span>
  </li>
);
