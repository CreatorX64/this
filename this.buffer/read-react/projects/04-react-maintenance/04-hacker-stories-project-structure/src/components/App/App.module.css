/* If using CSS-in-JS, where no CSS file is needed, one could still have a
separate style.js file for all the styled components */

.container {
  min-height: 100vh;
  padding: 20px;
  color: #171212;
  background: #83a4d4; /* fallback */
  background: linear-gradient(to left, #b6fbff, #83a4d4);
}

.headlinePrimary {
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
}
