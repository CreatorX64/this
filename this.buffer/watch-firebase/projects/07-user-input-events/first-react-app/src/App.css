.App {
  text-align: center;
}

.btn {
  margin: 0 10px;
}
