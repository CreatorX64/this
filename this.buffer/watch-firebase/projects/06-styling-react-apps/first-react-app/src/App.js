import { useState } from "react";
import "./App.css";
import Title from "./components/Title";
import Modal from "./components/Modal";
import EventList from "./components/EventList";

const App = () => {
  const [showModal, setShowModal] = useState(false);
  const [showEvents, setShowEvents] = useState(true);
  const [events, setEvents] = useState([
    { title: "Mario's birthday bash", id: 1 },
    { title: "Bowser's live stream", id: 2 },
    { title: "Race on moo moo farm", id: 3 }
  ]);

  const handleDelete = (id) => {
    setEvents((prevEvents) => prevEvents.filter((event) => event.id !== id));
  };

  const handleClose = () => {
    setShowModal(false);
  };

  const subtitle = "All the latest events in Marioland";

  return (
    <div className="App">
      <Title title="Events In Your Area" subtitle={subtitle} />

      <div>
        {showEvents && (
          <button className="btn" onClick={() => setShowEvents(false)}>
            Hide events
          </button>
        )}
        {!showEvents && (
          <button className="btn" onClick={() => setShowEvents(true)}>
            Show events
          </button>
        )}

        <button onClick={() => setShowModal(true)}>Show modal</button>
      </div>

      {showEvents && <EventList events={events} onDelete={handleDelete} />}

      <Modal isOpen={showModal} onClose={handleClose} isSalesModal={true}>
        <h2>Terms and Conditions</h2>
        <p>
          Lorem ipsum dolor sit, amet consectetur adipisicing elit. A quis eos
          eius, deserunt similique laborum quaerat numquam impedit commodi
          corrupti dolorum neque non dolor. Unde sunt recusandae optio magnam
          deleniti?
        </p>
      </Modal>
    </div>
  );
};

export default App;
