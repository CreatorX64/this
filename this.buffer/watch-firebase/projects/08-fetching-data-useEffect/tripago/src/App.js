import { useState } from "react";
import "./App.css";
import TripList from "./components/TripList";

const App = () => {
  const [showTrips, setShowTrips] = useState(true);

  return (
    <div className="App">
      <div>
        <button onClick={() => setShowTrips(false)}>Hide trips</button>
      </div>
      {showTrips && <TripList />}
    </div>
  );
};

export default App;
