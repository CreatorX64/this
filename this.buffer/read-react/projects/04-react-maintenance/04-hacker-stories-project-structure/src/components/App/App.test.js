// React Testing Library with Jest is the most popular library combination for
// React testing. RTL provides relevant testing tools, while Jest has a general
// testing framework for test suites, test cases, assertions, and mocking
// capabilities. If you need an alternative to RTL, consider trying Enzyme by Airbnb.

import axios from "axios";
import React from "react";
import { render, screen, fireEvent, act } from "@testing-library/react";
import { storyOne, storyTwo, stories } from "../../dummy-data";
import App, { storiesReducer } from "./index";

jest.mock("axios");

//-- The “describe” block is our "test suite", and the “test” blocks are our
// "test cases". Large subjects like functions or components often require
// multiple test cases, so it makes sense to use them with test suites

describe("something truthy and falsy", () => {
  test("true to be true", () => {
    // "Test assertions" 👇
    // expect(true).toBe(true);
    expect(true).toBeTruthy();
  });

  test("false to be false", () => {
    // "Test assertions" 👇
    // expect(false).toBe(false);
    // expect(false).toBe(true);
    expect(false).toBeFalsy();
  });

  // The “test” block can also be written as an “it” block. The blocks have
  // the same purpose

  // it("true to be true", () => {
  //   expect(true).toBe(true);
  // });

  // it("false to be false", () => {
  //   expect(false).toBe(false);
  // });
});

//-- Test cases can be used without test suites

// test("true to be true", () => {
//   expect(true).toBe(true);
// });

// test("false to be false", () => {
//   expect(false).toBe(false);
// });

//-- Unit testing: Functions

describe("storiesReducer", () => {
  test("removes a story from all stories", () => {
    const action = { type: "REMOVE_STORY", payload: storyOne };
    const state = { data: stories, isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [storyTwo],
      isLoading: false,
      isError: false
    };

    // Also note that there is a toEqual function which works slightly
    // different than toStrictEqual: https://twitter.com/rwieruch/status/1260866850080067584
    expect(newState).toStrictEqual(expectedState);
  });

  test("sets fetch init state", () => {
    const action = { type: "STORIES_FETCH_INIT" };
    const state = { data: [], isLoading: false, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [],
      isLoading: true,
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test("sets fetch success state", () => {
    const action = { type: "STORIES_FETCH_SUCCESS", payload: stories };
    const state = { data: [], isLoading: true, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: stories,
      isLoading: false,
      isError: false
    };

    expect(newState).toStrictEqual(expectedState);
  });

  test("sets fetch failure state", () => {
    const action = { type: "STORIES_FETCH_FAILURE" };
    const state = { data: [], isLoading: true, isError: false };

    const newState = storiesReducer(state, action);

    const expectedState = {
      data: [],
      isLoading: false,
      isError: true
    };

    expect(newState).toStrictEqual(expectedState);
  });
});

//-- Integration testing: Components: React Testing Library adheres to a single
// core philosophy: instead of testing implementation details of React components,
// it tests how users interact with the application and if it works as expected.
// This becomes especially powerful for integration tests

describe("App", () => {
  test("succeeds fetching data", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories
      }
    });

    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    // Use getBy* to check if an element is in the document, using getBy* queries
    // when asserting if element is present, so if the element is not found the
    // error thrown will offer better info than asserting with other queries
    // which will not throw an error
    expect(screen.getByText(/Loading/)).toBeInTheDocument();

    // Even though our Promise above resolves immediately, React updates the UI
    // async. That's why we have to use act() before we assert that our results
    // are rendered. We pass in a dummy Promise in act() to make sure that all
    // promises (aka microtasks in the microtask queue) are resolved & UI is
    // updated before we assert anything regarding the updated UI
    await act(() => Promise.resolve());

    // Use queryBy* to check if an element is NOT in the document, using
    // queryBy* queries when asserting if element is not present, so the test
    // doesn't fail immediately when the element is not found and the
    // assertion can be executed properly
    expect(screen.queryByText(/Loading/)).toBeNull();

    // Assert whether or not fetched data gets rendered as expected
    expect(screen.getByText("React")).toBeInTheDocument();
    expect(screen.getByText("Redux")).toBeInTheDocument();
    expect(screen.getAllByText("Dismiss").length).toBe(2);
  });

  test("fails fetching data", async () => {
    const promise = Promise.reject();

    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    expect(screen.getByText(/Loading/)).toBeInTheDocument();

    try {
      await act(() => Promise.resolve());
    } catch (error) {
      // Ignore errors
    } finally {
      expect(screen.queryByText(/Loading/)).toBeNull();
      expect(screen.getByText(/went wrong/)).toBeInTheDocument();
    }
  });

  test("removes a story", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories
      }
    });

    axios.get.mockImplementationOnce(() => promise);

    render(<App />);

    await act(() => Promise.resolve());

    expect(screen.getAllByText("Dismiss").length).toBe(2);
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();

    // Since the item with “Jordan Walke” is the first rendered item in the
    // list, it gets removed if we click the first “Dismiss” button
    fireEvent.click(screen.getAllByText("Dismiss")[0]);

    expect(screen.getAllByText("Dismiss").length).toBe(1);
    expect(screen.queryByText("Jordan Walke")).toBeNull();
  });

  test("searches for specific stories", async () => {
    const reactStoriesPromise = Promise.resolve({
      data: {
        hits: stories
      }
    });

    const anotherStory = {
      title: "JavaScript",
      url: "https://en.wikipedia.org/wiki/JavaScript",
      author: "Brendan Eich",
      num_comments: 15,
      points: 10,
      objectID: 3
    };

    const javaScriptStoriesPromise = Promise.resolve({
      data: {
        hits: [anotherStory]
      }
    });

    // Instead of mocking the request once with Jest, now we mock multiple requests
    axios.get.mockImplementation((url) => {
      if (url.includes("React")) {
        return reactStoriesPromise;
      }

      if (url.includes("JavaScript")) {
        return javaScriptStoriesPromise;
      }

      throw Error();
    });

    // Initial render
    render(<App />);

    // First data fetching

    await act(() => Promise.resolve());

    expect(screen.getByDisplayValue("React")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("JavaScript")).toBeNull();
    expect(screen.getByText("Jordan Walke")).toBeInTheDocument();
    expect(screen.getByText("Dan Abramov, Andrew Clark")).toBeInTheDocument();
    expect(screen.queryByText("Branden Eich")).toBeNull();

    // User interaction: Search

    fireEvent.change(screen.getByDisplayValue("React"), {
      target: {
        value: "JavaScript"
      }
    });

    expect(screen.queryByDisplayValue("React")).toBeNull();
    expect(screen.getByDisplayValue("JavaScript")).toBeInTheDocument();

    fireEvent.submit(screen.getByText("Submit"));

    // Second data fetching

    await act(() => Promise.resolve());

    expect(screen.queryByText("Jordan Walke")).toBeNull();
    expect(screen.queryByText("Dan Abramov, Andrew Clark")).toBeNull();
    expect(screen.getByText("Brendan Eich")).toBeInTheDocument();
  });

  test("renders snapshot", async () => {
    const promise = Promise.resolve({
      data: {
        hits: stories
      }
    });

    axios.get.mockImplementationOnce(() => promise);

    const { container } = render(<App />);

    await act(() => Promise.resolve());

    expect(container).toMatchSnapshot();
  });
});
