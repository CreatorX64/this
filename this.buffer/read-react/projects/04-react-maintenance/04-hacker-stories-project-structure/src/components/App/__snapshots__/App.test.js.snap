// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders snapshot 1`] = `
<div>
  <div
    class="container"
  >
    <h1
      class="headlinePrimary"
    >
      My Hacker Stories
    </h1>
    <form
      class="searchForm"
    >
      <label
        class="label"
        for="search"
      >
        <strong>
          Search:
        </strong>
      </label>
      Â 
      <input
        class="input"
        id="search"
        type="text"
        value="JavaScript"
      />
      <button
        class="button button-large"
        type="submit"
      >
        Submit
      </button>
    </form>
    <ul>
      <li
        class="item"
      >
        <span
          style="width: 40%;"
        >
          <a
            href="https://reactjs.org"
          >
            React
          </a>
        </span>
        <span
          style="width: 30%;"
        >
          Jordan Walke
        </span>
        <span
          style="width: 10%;"
        >
          3
        </span>
        <span
          style="width: 10%;"
        >
          4
        </span>
        <span
          style="width: 10%;"
        >
          <button
            class="button button-small"
            type="button"
          >
            Dismiss
          </button>
        </span>
      </li>
      <li
        class="item"
      >
        <span
          style="width: 40%;"
        >
          <a
            href="https://redux.js.org"
          >
            Redux
          </a>
        </span>
        <span
          style="width: 30%;"
        >
          Dan Abramov, Andrew Clark
        </span>
        <span
          style="width: 10%;"
        >
          2
        </span>
        <span
          style="width: 10%;"
        >
          5
        </span>
        <span
          style="width: 10%;"
        >
          <button
            class="button button-small"
            type="button"
          >
            Dismiss
          </button>
        </span>
      </li>
    </ul>
  </div>
</div>
`;
